lib/src/main/java/com/diffplug/spotless/npm/ShadowCopy.java

```diff
@@ -49,70 +48,63 @@ public ShadowCopy(@Nonnull File shadowCopyRoot) {
 	}
 
 	public void addEntry(String key, File orig) {
-		// prevent concurrent adding of entry with same key
-		if (!reserveSubFolder(key)) {
-			logger.debug("Shadow copy entry already in progress: {}. Awaiting finalization.", key);
+		File target = entry(key, orig.getName());
+		if (target.exists()) {
+			logger.debug("Shadow copy entry already exists, not overwriting: {}", key);
+		} else {
 			try {
-				NpmResourceHelper.awaitFileDeleted(markerFilePath(key).toFile(), Duration.ofSeconds(120));
-			} catch (TimeoutException e) {
-				throw new RuntimeException(e);
+				storeEntry(key, orig, target);
+			} catch (Throwable ex) {
+				// Log but don't fail
+				logger.error("Unable to store cache entry for {}", key, ex);
 			}
 		}
+	}
+
+	private void storeEntry(String key, File orig, File target) throws IOException {
+		// Create a temp directory in the same directory as target
+		Files.createDirectories(target.toPath().getParent());
+		Path tempDirectory = Files.createTempDirectory(target.toPath().getParent(), key);
+
 		try {
-			storeEntry(key, orig);
+			// Copy orig to temp dir
+			Files.walkFileTree(orig.toPath(), new CopyDirectoryRecursively(tempDirectory, orig.toPath()));
+
+			try {
+				// Atomically rename the completed cache entry into place
+				Files.move(tempDirectory, target.toPath(), StandardCopyOption.ATOMIC_MOVE);
```

Comment:
should we catch and have an alternative for fs's that dont support atomic move ?